File: problems/21_merge_two_sorted_lists/merge_two_sorted_list.rs
/// Definition of a singly-linked list node.
/// Uses `Option<Box<ListNode>>` for ownership-safe, in-place manipulation.
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    /// Creates a new list node with given value.
    #[inline]
    pub fn new(val: i32) -> Self {
        ListNode { val, next: None }
    }
}

/// Solution struct to group our method for LeetCode.
pub struct Solution;

impl Solution {
    /// Merge two sorted linked lists `l1` and `l2` in-place and return the head of the merged list.
    ///
    /// # Complexity
    /// - Time: O(n + m) where n and m are the lengths of `l1` and `l2`.
    /// - Space: O(1) extra, reuses existing nodes.
    pub fn merge_two_lists(
        mut l1: Option<Box<ListNode>>,
        mut l2: Option<Box<ListNode>>,
    ) -> Option<Box<ListNode>> {
        // Dummy node simplifies edge cases (empty inputs, head insertion)
        let mut dummy = Box::new(ListNode::new(0));
        let mut tail = &mut dummy;

        // Continue until one list is exhausted
        while l1.is_some() && l2.is_some() {
            // Compare head values safely with as_ref().unwrap().val
            if l1.as_ref().unwrap().val < l2.as_ref().unwrap().val {
                // Detach node from l1
                let mut node = l1.unwrap();
                l1 = node.next.take();
                tail.next = Some(node);
            } else {
                // Detach node from l2
                let mut node = l2.unwrap();
                l2 = node.next.take();
                tail.next = Some(node);
            }
            // Advance the tail
            tail = tail.next.as_mut().unwrap();
        }

        // Attach any remaining nodes (only one of l1/l2 is non-empty)
        tail.next = if l1.is_some() { l1 } else { l2 };

        // Skip dummy and return the real merged head
        dummy.next
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    /// Builds a linked list from a slice, returning its head.
    fn build_list(vals: &[i32]) -> Option<Box<ListNode>> {
        let mut dummy = Box::new(ListNode::new(0));
        let mut tail = &mut dummy;
        for &v in vals {
            tail.next = Some(Box::new(ListNode::new(v)));
            tail = tail.next.as_mut().unwrap();
        }
        dummy.next
    }

    /// Converts a linked list to a Vec<i32> for easy comparison.
    fn list_to_vec(mut head: Option<Box<ListNode>>) -> Vec<i32> {
        let mut result = vec![];
        while let Some(node) = head {
            result.push(node.val);
            head = node.next;
        }
        result
    }

    #[test]
    fn test_merge_two_lists() {
        let cases = vec![
            (vec![], vec![], vec![]),
            (vec![], vec![0], vec![0]),
            (vec![1, 2, 4], vec![1, 3, 4], vec![1, 1, 2, 3, 4, 4]),
        ];

        for (a, b, expected) in cases {
            let l1 = build_list(&a);
            let l2 = build_list(&b);
            let merged = Solution::merge_two_lists(l1, l2);
            assert_eq!(list_to_vec(merged), expected);
        }
    }
}

File: problems/21_merge_two_sorted_lists/merge_two_sorted_list.java
// LeetCode #21: Merge Two Sorted Lists
// Merge two sorted linked lists into a single sorted linked list in-place.

/**
 * Definition for singly-linked list.
 */
class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}

public class MergeTwoSortedList {

    /**
     * Merges two sorted linked lists `list1` and `list2` in-place.
     * 
     * @param list1 head of the first sorted linked list
     * @param list2 head of the second sorted linked list
     * @return head of the merged sorted linked list
     *
     * Time complexity: O(n + m), where n and m are lengths of `list1` and `list2`.
     * Space complexity: O(1) extra (in-place).
     */
    public static ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        ListNode dummy = new ListNode(-1);
        ListNode tail = dummy;

        while (list1 != null && list2 != null) {
            if (list1.val < list2.val) {
                tail.next = list1;
                list1 = list1.next;
            } else {
                tail.next = list2;
                list2 = list2.next;
            }
            tail = tail.next;
        }

        tail.next = (list1 != null) ? list1 : list2;

        return dummy.next;
    }

    // Helper method to create linked list from array
    public static ListNode buildList(int[] values) {
        ListNode dummy = new ListNode(0);
        ListNode current = dummy;
        for (int val : values) {
            current.next = new ListNode(val);
            current = current.next;
        }
        return dummy.next;
    }

    // Helper method to print linked list
    public static void printList(ListNode head) {
        while (head != null) {
            System.out.print(head.val + " -> ");
            head = head.next;
        }
        System.out.println("null");
    }

    // Example usage and simple tests
    public static void main(String[] args) {
        int[][][] tests = {
            {{}, {}},
            {{}, {0}},
            {{1, 2, 4}, {1, 3, 4}}
        };

        for (int[][] test : tests) {
            ListNode l1 = buildList(test[0]);
            ListNode l2 = buildList(test[1]);
            ListNode merged = mergeTwoLists(l1, l2);
            printList(merged);
        }
    }
}
